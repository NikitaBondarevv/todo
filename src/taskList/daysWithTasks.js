import { Component } from 'react'

import { TaskList } from '.'
import { daysOfTheWeek } from '../helpers/constans'
import { Tabs } from './tabs'
import { Tab } from './tabs/tab'
import { getTasks } from '../contracts/getTasks'

export class DaysWithTasks extends Component {
  state = {
    days: []
  }

  async getTasks() {
    this.setState({ days: await getTasks() })
  }

  componentDidMount() {
    this.getTasks()
  }

  getDayNow = () => {
    let dayNow = new Date().getDay()

    if (dayNow === 0) {
      return dayNow += 6
    }

    return dayNow - 1
  }

  render() {
    const { days } = this.state

    return (
      <Tabs selectedIndex={this.getDayNow()}>
        {
          days.map((day, index) => (
            <Tab key={index} title={daysOfTheWeek[index]}>
              <TaskList tasks={day} />
            </Tab>
          ))
        }
      </Tabs>
    )
  }
}
