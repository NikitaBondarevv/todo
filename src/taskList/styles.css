@font-face {
  font-family: "MyriadProBold";
  src: url(../../fonts/MyriadProBold.ttf);
}

.tasksList {
  margin: 25px 0;
  padding: 0 15px 0;
  font-size: 24px;
  line-height: 24px;
  color: #fff;
  list-style: none;
}

.tasksList li {
  display: flex;
  align-items: center;
  margin: 0 0 15px;
  column-gap: 15px;
}

.tasksList span {
  height: 35px;
  min-width: 35px;
}

.tasksList p {
  margin: 0;
}

.done {
  text-decoration: line-through;
}

.taskInProgress {
  text-decoration: underline;
}

.waiting {
  background: url(./images/waiting.png) no-repeat;
}

.done {
  background: url(./images/done.png) no-repeat;
}

.progress {
  background: url(./images/in-progress.png) no-repeat;
}

.tasksList a {
  text-indent: -9999px;
}

.tasks a {
  display: block;
  background: url(./images/hovering.png) no-repeat;
  width: 38px;
  height: 38px;
  cursor: pointer;
}

.tasks .inProgress {
  background-position: -53px 0;
}

.tasks .delete {
  background-position: -105px 0;
}

.buttons {
  display: flex;
  column-gap: 10px;
}

.content,
.activeContent {
  width: 837px;
  margin: 0 auto;
  background: rgba(0, 0, 255, .3);
  border: 1px solid #fff;
  box-shadow: 0 0 5px 1px rgb(0 0 0 / 50%);
  padding: 20px;
}

.tabsNav {
  transform: translate(53.5px, -78px);
}

.list {
  display: flex;
  list-style: none;
  margin: 0;
  padding: 0;
  font-size: 18px;
  font-family: "MyriadProBold";
  line-height: 24px;
}

.list a {
  display: block;
  text-decoration: none;
  color: #fff;
  text-align: center;
  padding: 16px 22px;
  background: rgba(0, 0, 255, .3);
  box-shadow: 0 0 5px 1px rgb(0 0 0 / 50%);
  border: 1px solid #fff;
}

.list a:hover {
  text-decoration: underline;
}

.active a {
  box-shadow: 0 -2px 2px rgb(0 0 0 / 50%);
  font-weight: bold;
  background: #303994
}

.activeContent {
  background: #303994
}

.addTask {
  background: #fff;
  width: 244px;
  font-size: 22px;
  height: 55px;
  line-height: 55px;
  text-align: center;
  text-decoration: underline;
  box-shadow: 0px 2px 5px #000;
}

.editableText {
  height: 45px;
  width: 234px;
  font-size: 22px;
}
