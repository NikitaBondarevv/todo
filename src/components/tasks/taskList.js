import { useState, useEffect } from 'react'
import PropTypes from 'prop-types'

import { UpdateCreateTask } from '../updateCreateTask'
import { Task } from './task'
import styles from './styles.css'

export const TaskList = ({ tasks, getTasks, activeTabIndex }) => {
  const [todos, setTodos] = useState(tasks)

  useEffect(() => {
    setTodos(tasks)
  }, [tasks])

  return (
    <>
      <div>
        <ol className={styles.tasksList}>
          {
            todos.map((task, index) => <Task key={index} data={task} getTasks={getTasks} />)
          }
        </ol>
        <a href="#" className={styles.addTask}>Add new task</a>
      </div>
      <UpdateCreateTask text="New Task" getTasks={getTasks} activeTabIndex={activeTabIndex} />
    </>
  )
}

TaskList.defaultProps = {
  tasks: [],
  getTasks: () => { },
  activeTabIndex: 0
}

TaskList.propTypes = {
  tasks: PropTypes.array,
  getTasks: PropTypes.func,
  activeTabIndex: PropTypes.number
}
