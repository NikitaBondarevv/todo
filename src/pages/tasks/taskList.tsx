import { useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'

import { Task } from './task'
import { TTaskListProps } from './types'
import styles from './styles.css'

export const TaskList = ({ tasks, getTasks, activeTabIndex }: TTaskListProps) => {
  const [todos, setTodos] = useState(tasks)

  useEffect(() => {
    setTodos(tasks)
  }, [tasks])

  return (
    <div>
      <ol className={styles.tasksList}>
        {
          todos.map((task, index) => <Task key={index} data={task} getTasks={getTasks} />)
        }
      </ol>
      <Link to={`/task/${activeTabIndex}`} className={styles.addTask}>Add new task</Link>
    </div>
  )
}

TaskList.defaultProps = {
  tasks: [],
  activeTabIndex: 0,
  getTasks: () => {}
}

TaskList.propTypes = {
  tasks: PropTypes.array,
  getTasks: PropTypes.func,
  activeTabIndex: PropTypes.number
}
