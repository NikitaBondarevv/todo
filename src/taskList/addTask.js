import { Component } from 'react'
import styles from './styles.css'

export class AddTask extends Component {
  state = {
    hidden: true,
    value: ''
  }

  setValue = ({ target: { value } }) => {
    this.setState({ value });
  }

  handleBlur = (e) => {
    e.preventDefault();

    this.setState({ hidden: true });
  }

  showInput = () => {
    this.setState({ hidden: false });
  }

  render() {
    const { value, hidden } = this.state;

    return (
      <div className={styles.addTask}>
        {
          hidden ? <span onClick={this.showInput}>{'Add new task'}</span> :
            <form onSubmit={this.handleBlur}>
              <input className={styles.editableText} name="text" value={value} onChange={this.setValue} onBlur={this.handleBlur} autoFocus />
            </form>
        }
      </div>
    )
  }
}