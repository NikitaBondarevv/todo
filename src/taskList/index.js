import { Component } from 'react'

import styles from './styles.css'

const Tasks = ({ id, completed, title, buttons }) => (
  <li key={id} className={completed ? styles.completed : styles.tasks}>
    <span className={completed ? styles.done : completed === undefined ? styles.progress : styles.waiting}></span>
    <p className={completed === undefined ? styles.taskInProgress : ''} >{`${id}. ${title}`}</p>
    {buttons}
  </li>
)

export class TaskList extends Component {
  state = {
    todos: [],
    value: ''
  }

  originTodos = []

  async getTasks() {
    const response = await fetch('https://jsonplaceholder.typicode.com/todos')
    const todos = await response.json()

    this.originTodos = todos

    this.setState({ todos: this.originTodos })
  }

  componentDidMount() {
    this.getTasks()
  }

  completeTask = (index) => {
    this.state.todos[index].completed = true

    this.setState({ todos: this.state.todos })
  }

  setInProgressTask = (index) => {
    this.state.todos[index].completed = undefined

    this.setState({ todos: this.state.todos })
  }

  deleteTask = (id) => {
    this.setState({
      todos: this.state.todos.filter(todo => todo.id !== id)
    })
  }

  taskSearch = ({ target: { value } }) => {
    this.setState({
      value,
      todos: this.originTodos
    })

    if (value.length > 1) {
      this.setState({ todos: this.originTodos.filter(todo => todo.title.includes(value)) })
    }
  }

  render() {
    const { todos, value } = this.state

    return (
      <>
        <input value={value} onChange={this.taskSearch} className={styles.search} placeholder="SEARCH" />
        <ul className={styles.tasksList}>
          {
            todos.map((todo, index) =>
              <Tasks key={index} id={index + 1} completed={todo.completed} title={todo.title} buttons={
                <div className={styles.buttons} >
                  <a onClick={() => this.completeTask(index)}></a>
                  {todo.completed === undefined ? '' : <a onClick={() => this.setInProgressTask(index)} className={styles.inProgress}></a>}
                  <a onClick={() => this.deleteTask(todo.id)} className={styles.delete}></a>
                </div>
              } />
            )
          }
        </ul>
      </>
    )
  }
}
