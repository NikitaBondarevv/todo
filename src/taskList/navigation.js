import styles from './styles.css'

import PropTypes from 'prop-types'

export const Navigation = ({ titles, setActiveTab, activeTabIndex }) => {
  return (
    <nav className={styles.tabsNav} >
      <ul className={styles.list} >
        {titles.map((title, index) =>
          <li key={index} className={activeTabIndex === index ? styles.active : ''} ><a href="#" onClick={() => setActiveTab(index)}>{title}</a></li>
        )}
      </ul>
    </nav>
  )
}

Navigation.defaultProps = {
  titles: [],
  activeTabIndex: 0,
  setActiveTab: function () {},
}

Navigation.propTypes = {
  titles: PropTypes.array,
  activeTabIndex: PropTypes.number,
  setActiveTab: PropTypes.func,
}
