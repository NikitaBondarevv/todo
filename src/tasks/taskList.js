import { Component } from 'react'
import PropTypes from 'prop-types'

import { Task } from './task'
import styles from './styles.css'

export class TaskList extends Component {
  static defaultProps = {
    tasks: []
  }

  static propTypes = {
    tasks: PropTypes.array
  }

  state = {
    todos: this.props.tasks,
    hidden: true,
    value: ''
  }

  componentDidUpdate(prevProps) {
    if (this.props.tasks !== prevProps.tasks) {

      this.setState({ todos: this.props.tasks })
    }
  }

  setValue = ({ target: { value } }) => {
    this.setState({ value })
  }

  showInput = () => {
    this.setState({ hidden: false })
  }

  randomInteger = () => {
    const rand = 1 + Math.random() * (100 + 1 - 1);

    return Math.floor(rand);
  }

  addTaskAndHandleBlur = (e) => {
    e.preventDefault()

    if (this.state.value.length > 2) {
      this.state.todos.push({
        title: this.state.value,
        id: this.randomInteger(),
        done: false
      })
    }

    this.setState({
      todos: this.state.todos,
      hidden: true,
      value: ''
    })
  }

  render() {
    const { todos, value, hidden } = this.state

    return (
      <div>
        <ul className={styles.tasksList}>
          {
            todos.map((task, index) => <Task key={index} index={index + 1} data={task} getTasks={this.props.getTasks} />)
          }
        </ul>
        <div className={styles.addTask}>
          {
            hidden ? <span onClick={this.showInput}>Add new task</span> :
              <form onSubmit={this.addTaskAndHandleBlur}>
                <input className={styles.editableText}
                  name="text"
                  value={value}
                  onChange={this.setValue}
                  onBlur={this.addTaskAndHandleBlur}
                  autoFocus
                />
              </form>
          }
        </div>
      </div>
    )
  }
}
