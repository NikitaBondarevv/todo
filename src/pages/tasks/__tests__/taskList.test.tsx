import { render } from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'

import { TaskList } from '../taskList'

describe('<TaskList />', () => {
  test('should match snapshot', () => {
    const { asFragment } = render(
      <BrowserRouter>
        <TaskList tasks={[{ title: 'test', day: '0', id: '0', done: false }]} getTasks={() => { }} activeTabIndex={0} />
      </BrowserRouter>
    )

    expect(asFragment()).toMatchSnapshot()
  })
})
