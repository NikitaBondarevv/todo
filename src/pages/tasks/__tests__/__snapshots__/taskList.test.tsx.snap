// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TaskList /> should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <ol
      class="tasksList"
    >
      <li
        class="waiting updateTask"
      >
        Â 
        <a
          href="/task/0/0"
        >
          test
        </a>
        <div
          class="buttons"
        >
          <button
            type="button"
          />
          <button
            class="inProgress"
            type="button"
          />
          <button
            class="delete"
            type="button"
          />
        </div>
      </li>
    </ol>
    <a
      class="addTask"
      href="/task/0"
    >
      Add new task
    </a>
  </div>
</DocumentFragment>
`;
